tools/apprunner-common/src/main/java/org/projectnessie/quarkus/runner/ProcessHandler.java

```diff
@@ -0,0 +1,273 @@
+   * @throws TimeoutException if the Quarkus process did not write the listen URL to stdout.
+   */
+  public String getListenUrl() throws InterruptedException, TimeoutException {
+    return listenUrlWaiter.getListenUrl();
+  }
+
+  /**
+   * Stops the process.
+   * <p>Tries to gracefully stop the process via a {@code SIGTERM}. If the process is still alive
+   * after {@link #setTimeStopMillis(long)}, which defaults to {@value #MILLIS_TO_STOP} ms,
+   * the process will be killed with a {@code SIGKILL}.</p>
+   */
+  public void stop() {
+    if (process == null) {
+      throw new IllegalStateException("No process started");
+    }
+
+    doStop();
+
+    watchdogExitGrace();
+  }
+
+  private void shutdownHandler() {
+    doStop();
+  }
+
+  private void doStop() {
+    if (stopped.compareAndSet(false, true)) {
+      try {
+        listenUrlWaiter.cancel();
+        process.destroy();
+        try {
+          if (!process.waitFor(timeStopMillis, TimeUnit.MILLISECONDS)) {
+            process.destroyForcibly();
+          }
+        } catch (InterruptedException e) {
+          process.destroyForcibly();
+          Thread.currentThread().interrupt();
+        }
+        watchdogExecutor.shutdown();
+      } finally {
+        if (Thread.currentThread() != shutdownHook) {
+          // Don't remove the shutdown-hook if we're running in the shutdown-hook
+          Runtime.getRuntime().removeShutdownHook(shutdownHook);
+        }
+      }
+    }
+  }
+
+  void watchdogExitGrace() {
+    try {
+      // Give the watchdog task/thread some time to finish its work
+      watchdogFuture.get(timeStopMillis, TimeUnit.MILLISECONDS);
+    } catch (ExecutionException e) {
+      throw new RuntimeException("ProcessHandler's watchdog thread failed.", e);
+    } catch (TimeoutException e) {
+      throw new IllegalStateException("ProcessHandler's watchdog thread failed to finish in time.");
+    } catch (InterruptedException e) {
+      process.destroyForcibly();
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  public boolean isAlive() {
+    return exitCode.get() == RUNNING;
+  }
+
+  /**
+   * Retrieves the exit-code of the process, if it terminated or throws a
+   * {@link IllegalThreadStateException} if it is still alive.
+   *
+   * @return the exit code of the process
+   * @throws IllegalThreadStateException if the process is still alive
+   */
+  public int getExitCode() throws IllegalThreadStateException {
+    if (isAlive()) {
+      throw new IllegalThreadStateException();
+    }
+    return exitCode.get();
+  }
+
+  private Object watchdog() throws IOException {
+    try (InputStream out = process.getInputStream(); InputStream err = process.getErrorStream()) {
+      InputBuffer stdout = new InputBuffer(out, listenUrlWaiter);
+      InputBuffer stderr = new InputBuffer(err, stderrTarget);
+      try {
+
+        while (true) {
+          boolean anyIo = stdout.io();
```

Comment:
I am not sure I get the point of checking for text logged to the screen, especially below where we enter a warm loop if no io and a hot loop if io. This while loop is unclear to me in general. can you say a few words?

Reply:
added some comment